### ISSUE TITLE
Support filter subject in mirror and sources for streams

### ISSUE THREAD

**Issue by dodomorandi** [nats-io/nats-server#4082]
## Feature Request

It is possible to setup consumers with `filter_subject` in order to have granular authorizations over `$JS.*.API.CONSUMER.CREATE`. However, the same is not possible when using stream sources and mirrors.

#### Use Case:

With this change it is possible to create streams and KV stores that only mirror/source a subset of the original one. Very useful for authorization purposes, very useful for the bandwidth.

#### Proposed Change:

The `FilterSubject` is already parsed and passed around, it is just a matter of putting it in the creation of the consumer.

#### Who Benefits From The Change(s)?

Everyone that has a "central" stream/kv store that needs to be replicated to leaf nodes with high granularity.

#### Alternative Approaches

N/A

CC @paolobarbolini

**Comment by derekcollison** [comment]
This is possible already today..

The StreamSource, which is how streams describe connections to mirrors and sources, has a FilterSubject on it already.


```go
// StreamSource dictates how streams can source from other streams.
type StreamSource struct {
	Name          string          `json:"name"`
	OptStartSeq   uint64          `json:"opt_start_seq,omitempty"`
	OptStartTime  *time.Time      `json:"opt_start_time,omitempty"`
	FilterSubject string          `json:"filter_subject,omitempty"`
	External      *ExternalStream `json:"external,omitempty"`
}

```

**Comment by dodomorandi** [comment]
Maybe my point is not clear enough -- check PR #4083 to check my point.

**Comment by derekcollison** [comment]
ok I think I might understand what you are referring to, the ability to control via subject permissions etc..

**Comment by dodomorandi** [comment]
@derekcollison sorry for my very short response, I did not have enough time to explain myself in a better way. Please, let me rephrase the issue in order to make things clearer.

When you want to perform data replication of streams using mirrors or specifying sources, it is already possible to specify the filter subject. However, the problem is that this field, available as `FilterSubject` inside the `StreamSource` struct, is not currently used during the creation of the relative consumer(s). As you correctly guessed, this is important when you want to use specific permissions in order to allow only `$JS.*.API.CONSUMER.CREATE.*.*.<filter>` for accounts and/or users. If the filter subject parameter is not _forwarded_ to the creation of the consumer, the JetStream API is called without the filter, therefore it is not possible to use it for authorization purposes.

Sorry again for not being clear enough yesterday :relaxed:.

**Comment by derekcollison** [comment]
Gotcha, this will be the case for 2.10 release.

**Comment by derekcollison** [comment]
Are these on other side of a leafnode?

**Comment by jnmoyne** [comment]
As I mentioned in https://github.com/nats-io/nats-server/pull/4083 it looks to me like the filter is already being included in the consumer creation request, just not when the struct is created but added a few lines below (e.g. https://github.com/M4SS-Code/nats-server/blob/e96ae0bf79a2a3737b00376636cd1d04194c244c/server/stream.go#L2277-L2280)

**Comment by dodomorandi** [comment]
Ouch, you are totally right. I totally missed the fact that the parameter was changed later on. Sorry guys, I hope I did not make you lose too much time. The reason behind this dummy thing is a simple feature I am going to propose in the next few days -- I (wrongly) noticed the missing `FilterSubject` parameter in the construction of the consumer while performing the other changes.

Sorry again and thank you for you efforts.
